WordMapping validation

OlafMeuwese opened the pull request:
Tests for 3 types of errors:
1. Invalid filenames (Error)
2. Invalid column numbers (Error)
3. Unmapped values (Warning)

Includes new tests which depends on a new study named Invalid Study

OlafMeuwese commented on tmtk/clinical/WordMapping.py:
```diff
@@ -31,11 +32,47 @@ def __init__(self, params=None):
+                absolute_filename = os.path.join(self.params.dirname, filename)
+                valid_filenames.append(absolute_filename)
+
+        dfs = {os.path.basename(filename): pd.read_table(filename) for filename in valid_filenames}
```
This seems less than ideal. Currently, the only link for the WordMapping to connect to the rest of the study is through the clinical.params that is provided.
Possible more information can be handed down, so the datafiles used by the ColumnMapping could be used instead of instantiating it here for testing.

jochemb commented on tmtk/clinical/WordMapping.py:
```diff
@@ -31,11 +32,47 @@ def __init__(self, params=None):
+        for filename, df in dfs.items():
+            column_number = df.shape[1]
+
+            rows = [row for index, row in self.df.iterrows() if row[0] == filename]
```
You should be able to use the indexes created: ``rows = self.df.loc[filename]``

jochemb commented on tmtk/clinical/WordMapping.py:
```diff
@@ -31,11 +32,47 @@ def __init__(self, params=None):
+            column_number = df.shape[1]
+
+            rows = [row for index, row in self.df.iterrows() if row[0] == filename]
+            columns = {row[1] for row in rows}
```
or direct way to get columns for a filename: `` list(self.df.loc[filename, self.df.columns[1]])``

jochemb replied:
Or if you like it dirty

`` columns = set(self.df.loc[filename].index) ``

jochemb commented on tmtk/clinical/WordMapping.py:
```diff
@@ -31,11 +32,47 @@ def __init__(self, params=None):
+
+            correct_columns = columns - out_of_bound
+            for column in correct_columns:
+                mapped_values = {row[2] for row in rows if row[1] == column}
```
``mapped_values = set(self.df.loc[(filename, column), self.df.columns[2]])`` 💯

jochemb commented on tmtk/clinical/WordMapping.py:
```diff
@@ -31,11 +32,47 @@ def __init__(self, params=None):
+            for column in correct_columns:
+                mapped_values = {row[2] for row in rows if row[1] == column}
+                index = column - 1
+                data_values = set(df.iloc[:, index].unique())
```
The Clinical object has a ``get_variable()`` method that wants a tuple of filename and column number. It returns a Variable object which does exactly some of the stuff you're doing here. It does not have an *mapped_but_not_present_values* property though, which would be nice to implement!

OlafMeuwese commented:
I reckon we should refactor this, like you mentioned. 
The refactor should then include:
1. Move to the Clinical
2. Using the dataframes, indices, methods etc. This will have the benefit of: 
     1. Instantiating dataframes only once
      1. More explicit logic

OlafMeuwese commented:
Refactored by:

1. Moved to Clinical as _validate_word_mappings
2. Using WordMapping, ColumnMapping & Variable Classes
3. Adding included_datafiles property to WordMapping
4. Added check to Variable class called word_mapped_not_present
